File: VB/SchedulingDemo.Wpf/ViewModels/OutlookInspired/PrintingSettingsWindowViewModel.vb
Imports System
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports System.IO
Imports System.Linq
Imports Microsoft.Win32
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.POCO
Imports DevExpress.Xpf.Scheduling
Imports DevExpress.Xpf.Scheduling.Reporting
Imports DevExpress.XtraReports.UI
Imports DevExpress.XtraScheduler.Reporting
Imports DevExpress.XtraPrinting.Native
Imports DevExpress.Xpf.Printing.Native

Namespace SchedulingDemo.ViewModels
    Public Class PrintingSettingsWindowViewModel
        Public Shared Function Create(ByVal scheduler As SchedulerControl, ByVal reportTemplateInfos As IList(Of ReportTemplateInfo)) As PrintingSettingsWindowViewModel
            Return ViewModelSource.Create(Function() New PrintingSettingsWindowViewModel(scheduler, reportTemplateInfos))
        End Function
        Private Shared Sub MoveSelectedItem(ByVal source As IList(Of ResourceItem), ByVal target As IList(Of ResourceItem), ByVal selectedItem As ResourceItem)
            Dim selectedItemIndex As Integer = source.IndexOf(selectedItem)
            If selectedItemIndex < 0 Then
                Return
            End If
            target.Add(source(selectedItemIndex))
            source.RemoveAt(selectedItemIndex)
        End Sub

        Private ReadOnly scheduler As SchedulerControl

        Protected Sub New(ByVal scheduler As SchedulerControl, ByVal reportTemplateInfos As IList(Of ReportTemplateInfo))
            Me.scheduler = scheduler
            AddHandler Me.scheduler.SchedulerPrintAdapter.ValidateResources, AddressOf OnValidateResources

            Me.ReportTemplateInfos = New ObservableCollection(Of ReportTemplateInfo)(reportTemplateInfos)

            If reportTemplateInfos.Count > 0 Then
                ActiveReportTemplateInfo = reportTemplateInfos(0)
            End If

            StartDate = scheduler.VisibleIntervals.First().Start
            EndDate = scheduler.VisibleIntervals.Last().End
            AvailableResources = New ObservableCollection(Of ResourceItem)(scheduler.VisibleResources)
            CustomResources = New ObservableCollection(Of ResourceItem)()
        End Sub

        Public Property UseSpecificResources() As Boolean
        Public Property StartDate() As Date
        Public Property EndDate() As Date
        Public Property ActiveReportTemplateInfo() As ReportTemplateInfo
        Public Property ReportTemplateInfos() As ObservableCollection(Of ReportTemplateInfo)
        Public Property AvailableResources() As ObservableCollection(Of ResourceItem)
        Public Property SelectedAvailableResource() As ResourceItem
        Public Property CustomResources() As ObservableCollection(Of ResourceItem)
        Public Property SelectedCustomResource() As ResourceItem
        Public Property PrintResources() As ObservableCollection(Of ResourceItem)

        Public Sub AddToCustomCollection()
            MoveSelectedItem(AvailableResources, CustomResources, SelectedAvailableResource)
        End Sub
        Public Sub RemoveFromCustomCollection()
            MoveSelectedItem(CustomResources, AvailableResources, SelectedCustomResource)
        End Sub
        Public Sub MoveUp()
            Dim index As Integer = CustomResources.IndexOf(SelectedCustomResource)
            If index <= 0 Then
                Return
            End If
            Dim item As ResourceItem = CustomResources(index)
            CustomResources.RemoveAt(index)
            CustomResources.Insert(index - 1, item)
        End Sub
        Public Sub MoveDown()
            Dim index As Integer = CustomResources.IndexOf(SelectedCustomResource)
            If (index < 0) OrElse (index >= CustomResources.Count - 1) Then
                Return
            End If
            Dim item As ResourceItem = CustomResources(index)
            CustomResources.RemoveAt(index)
            CustomResources.Insert(index + 1, item)
        End Sub
        Public Sub Preview()
            Dim report As XtraSchedulerReport = SchedulerReportFactory.Create(ActiveReportTemplateInfo.ReportTemplatePath, Me.scheduler, New DateTimeRange(StartDate, EndDate))
            report.PrintingSystem.ReplaceService(Of BackgroundPageBuildEngineStrategy)(New DispatcherPageBuildStrategy())
            report.ShowPreview()
        End Sub
        Public Sub OpenFileDialog()
            Dim dlg As New OpenFileDialog()
            dlg.CheckPathExists = True
            dlg.Filter = "Report template files (*.schrepx)|*.schrepx|All files (*.*)|*.*"

            Dim folderPath As String = Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments)
            dlg.InitialDirectory = (New DirectoryInfo(folderPath)).FullName

            Dim result? As Boolean = dlg.ShowDialog()
            If result Is Nothing OrElse result = False Then
                Return
            End If

            ActiveReportTemplateInfo = New ReportTemplateInfo(dlg.SafeFileName, dlg.FileName)
            ReportTemplateInfos.Add(ActiveReportTemplateInfo)
        End Sub
        Public Sub WindowClosing()
            RemoveHandler Me.scheduler.SchedulerPrintAdapter.ValidateResources, AddressOf OnValidateResources
        End Sub

        Protected Sub OnStartDateChanged()
            If EndDate < StartDate Then
                EndDate = StartDate
            End If
        End Sub
        Protected Sub OnEndDateChanged()
            If EndDate < StartDate Then
                StartDate = EndDate
            End If
        End Sub

        Private Sub OnValidateResources(ByVal sender As Object, ByVal e As ResourceItemsValidationEventArgs)
            Dim resources As IList(Of ResourceItem) = e.ResourceItems
            resources.Clear()
            If UseSpecificResources Then
                CustomResources.ForEach(Sub(x) resources.Add(x))
            Else
                AvailableResources.ForEach(Sub(x) resources.Add(x))
            End If
        End Sub
    End Class
End Namespace

File: CS/SchedulingDemo.Wpf/ViewModels/OutlookInspired/PrintingSettingsWindowViewModel.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.IO;
using System.Linq;
using Microsoft.Win32;
using DevExpress.Mvvm;
using DevExpress.Mvvm.POCO;
using DevExpress.Xpf.Scheduling;
using DevExpress.Xpf.Scheduling.Reporting;
using DevExpress.XtraReports.UI;
using DevExpress.XtraScheduler.Reporting;
using DevExpress.XtraPrinting.Native;
using DevExpress.Xpf.Printing.Native;

namespace SchedulingDemo.ViewModels {
    public class PrintingSettingsWindowViewModel {
        public static PrintingSettingsWindowViewModel Create(SchedulerControl scheduler, IList<ReportTemplateInfo> reportTemplateInfos) {
            return ViewModelSource.Create(() => new PrintingSettingsWindowViewModel(scheduler, reportTemplateInfos));
        }
        static void MoveSelectedItem(IList<ResourceItem> source, IList<ResourceItem> target, ResourceItem selectedItem) {
            int selectedItemIndex = source.IndexOf(selectedItem);
            if (selectedItemIndex < 0)
                return;
            target.Add(source[selectedItemIndex]);
            source.RemoveAt(selectedItemIndex);
        }

        readonly SchedulerControl scheduler;

        protected PrintingSettingsWindowViewModel(SchedulerControl scheduler, IList<ReportTemplateInfo> reportTemplateInfos) {
            this.scheduler = scheduler;
            this.scheduler.SchedulerPrintAdapter.ValidateResources += OnValidateResources;

            ReportTemplateInfos = new ObservableCollection<ReportTemplateInfo>(reportTemplateInfos);

            if (reportTemplateInfos.Count > 0)
                ActiveReportTemplateInfo = reportTemplateInfos[0];

            StartDate = scheduler.VisibleIntervals.First().Start;
            EndDate = scheduler.VisibleIntervals.Last().End;
            AvailableResources = new ObservableCollection<ResourceItem>(scheduler.VisibleResources);
            CustomResources = new ObservableCollection<ResourceItem>();
        }

        public bool UseSpecificResources { get; set; }
        public DateTime StartDate { get; set; }
        public DateTime EndDate { get; set; }
        public ReportTemplateInfo ActiveReportTemplateInfo { get; set; }
        public ObservableCollection<ReportTemplateInfo> ReportTemplateInfos { get; set; }
        public ObservableCollection<ResourceItem> AvailableResources { get; set; }
        public ResourceItem SelectedAvailableResource { get; set; }
        public ObservableCollection<ResourceItem> CustomResources { get; set; }
        public ResourceItem SelectedCustomResource { get; set; }
        public ObservableCollection<ResourceItem> PrintResources { get; set; }

        public void AddToCustomCollection() {
            MoveSelectedItem(AvailableResources, CustomResources, SelectedAvailableResource);
        }
        public void RemoveFromCustomCollection() {
            MoveSelectedItem(CustomResources, AvailableResources, SelectedCustomResource);
        }
        public void MoveUp() {
            int index = CustomResources.IndexOf(SelectedCustomResource);
            if (index <= 0)
                return;
            ResourceItem item = CustomResources[index];
            CustomResources.RemoveAt(index);
            CustomResources.Insert(index - 1, item);
        }
        public void MoveDown() {
            int index = CustomResources.IndexOf(SelectedCustomResource);
            if ((index < 0) || (index >= CustomResources.Count - 1))
                return;
            ResourceItem item = CustomResources[index];
            CustomResources.RemoveAt(index);
            CustomResources.Insert(index + 1, item);
        }
        public void Preview() {
            XtraSchedulerReport report = SchedulerReportFactory.Create(ActiveReportTemplateInfo.ReportTemplatePath, this.scheduler, new DateTimeRange(StartDate, EndDate));
            report.PrintingSystem.ReplaceService<BackgroundPageBuildEngineStrategy>(new DispatcherPageBuildStrategy());
            report.ShowPreview();
        }
        public void OpenFileDialog() {
            OpenFileDialog dlg = new OpenFileDialog();
            dlg.CheckPathExists = true;
            dlg.Filter = "Report template files (*.schrepx)|*.schrepx|All files (*.*)|*.*";

            string folderPath = Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments);
            dlg.InitialDirectory = new DirectoryInfo(folderPath).FullName;

            bool? result = dlg.ShowDialog();
            if (result == null || result == false)
                return;

            ActiveReportTemplateInfo = new ReportTemplateInfo(dlg.SafeFileName, dlg.FileName);
            ReportTemplateInfos.Add(ActiveReportTemplateInfo);
        }
        public void WindowClosing() {
            this.scheduler.SchedulerPrintAdapter.ValidateResources -= OnValidateResources;
        }

        protected void OnStartDateChanged() {
            if (EndDate < StartDate)
                EndDate = StartDate;
        }
        protected void OnEndDateChanged() {
            if (EndDate < StartDate)
                StartDate = EndDate;
        }

        void OnValidateResources(object sender, ResourceItemsValidationEventArgs e) {
            IList<ResourceItem> resources = e.ResourceItems;
            resources.Clear();
            if (UseSpecificResources)
                CustomResources.ForEach(x => resources.Add(x));
            else
                AvailableResources.ForEach(x => resources.Add(x));
        }
    }
}
